### PR TITLE
Use MaybeUninit in libcore

### PR DESCRIPTION
**Pull request by RalfJung**
All code by @japaric. This re-submits the second half of https://github.com/rust-lang/rust/pull/53508 (the first half is at https://github.com/rust-lang/rust/pull/54667). This is likely the one containing the perf regression.

### CODE REVIEW

**Comment by RalfJung** [comment]
@dtolnay  Eh, I submitted this. I do not think I should review it.^^

**Comment by dtolnay** [comment]
You already reviewed this code in #53508. Please clarify what remains to be reviewed.

Assigning the same reviewer as #54667:
r? @cramertj

**Comment by Valinora** [comment]
Perf results are in, this definitely looks to be the source of the regression.

**Comment by gnzlbg** [comment]
@RalfJung I don't know if this helps, but I don't see anything suspicious with this PR, so maybe the issue causing the regression is not here.

One "suspicious" thing is that the regression happens in `liballoc`, but `MaybeUninit` is declared in `libcore` and its methods are not `#[inline]`, which means that they cannot reliably be inlined into other crates without some form of LTO. 

If you look at the disassembly of these tests and see a `call` instruction pointing to the `MaybeUninit` methods in `libcore`, it might be worth testing whether making `MaybeUninit` methods `#[inline]` might fix this regression.

**Comment by RalfJung** [comment]
It's methods are generic though so `#[inline]` shouldn't make a difference, right?

> the regression happens in liballoc

Where are you getting that from?

Unfortunately I will not have much time to work on this this week.

**Comment by gnzlbg** [comment]
> It's methods are generic though so #[inline] shouldn't make a difference, right?

Indeed. The only way in which that could make a difference is if they call a non-generic function that is not `#[inline]`, but it appears that this is not the case. I'm out of ideas.

> Where are you getting that from?

I thought most of the changes here were in `liballoc`, but there are some changes in `libcore` as well. It might be worth it to figure out which of the changes (slice rotate, slice sort, ptr, liballoc's btree, etc.) introduces the regression.

**Comment by RalfJung** [comment]
I might be doing the wrong thing, but I failed to reproduce the perf regression... in fact, `issue-46449` got 12% *faster* with this PR in my measurement:
```
$ hyperfine -w 1 -m 3 -p 'rm target -rf' 'cargo +stage1 build --release' 'cargo +stage1.2 build --release'
Benchmark #1: cargo +stage1 build --release

  Time (mean ± σ):      3.532 s ±  0.014 s    [User: 6.576 s, System: 0.115 s]
 
  Range (min … max):    3.520 s …  3.548 s
 
Benchmark #2: cargo +stage1.2 build --release

  Time (mean ± σ):      3.166 s ±  0.059 s    [User: 6.180 s, System: 0.137 s]
 
  Range (min … max):    3.122 s …  3.233 s
 
Summary

  'cargo +stage1.2 build --release' ran
    1.12x faster than 'cargo +stage1 build --release'
```

**Comment by RalfJung** [comment]
I now restricted this PR to just the libcore changes. Let's see if we can get those in at least.

@bors try

**Comment by RalfJung** [comment]
@rust-timer build d4c1752c3f1a6dabb2fdf4a86ea986d5f55ecc46

**Comment by cramertj** [comment]
looks bad :(

**Comment by RalfJung** [comment]
Well that likely means at least the other half will look good.^^ I'll make a separate PR for it.

**Comment by RalfJung** [comment]
> in fact, issue-46449 got 12% faster with this PR in my measurement:

Ignore that, I had compiled one of the two compilers with debug assertions enabled....

**Comment by RalfJung** [comment]
I suspect it's caused by the changes in `ptr`. The changes in `swap` are optimized away by LLVM (in one monomorphization and as a free-standing function, I know that doesn't guarantee anything but I'd say it's a good sign).

However, `swap_nonoverlapping_bytes` seems to have IR changes. [Here are both functions so you can explore their IR](https://play.rust-lang.org/?gist=b182bfdc3791c1a8d543106ba85ed9c0&version=nightly&mode=release&edition=2015). Here's the IR diff:
```diff
-define void @_ZN10playground29swap_nonoverlapping_bytes_old17h49ccafcb349957d5E(i8* nocapture %x, i8* nocapture %y, i64 %len) unnamed_addr #0 {
+define void @_ZN10playground31swap_nonoverlapping_bytes_fixed17h9e199c2b3be909ffE(i8* nocapture %x, i8* nocapture %y, i64 %len) unnamed_addr #0 {
 start:
-  %t1 = alloca %"swap_nonoverlapping_bytes_old::UnalignedBlock", align 8
+  %t1.sroa.0 = alloca [4 x i64], align 8
+  %t.sroa.0 = alloca [4 x i64], align 32
   %0 = icmp ult i64 %len, 32
-  br i1 %0, label %bb3, label %bb4
+  br i1 %0, label %bb3, label %bb4.lr.ph
+
+bb4.lr.ph:                                        ; preds = %start
+  %t.sroa.0.0.sroa_cast13 = bitcast [4 x i64]* %t.sroa.0 to i8*
+  br label %bb4
 
 bb3:                                              ; preds = %bb4, %start
   %i.0.lcssa = phi i64 [ 0, %start ], [ %2, %bb4 ]
   %1 = icmp ult i64 %i.0.lcssa, %len
-  br i1 %1, label %bb11, label %bb18
+  br i1 %1, label %bb12, label %bb20
 
-bb4:                                              ; preds = %start, %bb4
-  %2 = phi i64 [ %5, %bb4 ], [ 32, %start ]
-  %i.016 = phi i64 [ %2, %bb4 ], [ 0, %start ]
-  %3 = getelementptr inbounds i8, i8* %x, i64 %i.016
-  %4 = getelementptr inbounds i8, i8* %y, i64 %i.016
-  %t.0..sroa_cast = bitcast i8* %3 to <4 x i64>*
-  %t.0.copyload = load <4 x i64>, <4 x i64>* %t.0..sroa_cast, align 1
+bb4:                                              ; preds = %bb4.lr.ph, %bb4
+  %2 = phi i64 [ 32, %bb4.lr.ph ], [ %5, %bb4 ]
+  %i.022 = phi i64 [ 0, %bb4.lr.ph ], [ %2, %bb4 ]
+  call void @llvm.lifetime.start.p0i8(i64 32, i8* nonnull %t.sroa.0.0.sroa_cast13)
+  %3 = getelementptr inbounds i8, i8* %x, i64 %i.022
+  %4 = getelementptr inbounds i8, i8* %y, i64 %i.022
+  call void @llvm.memcpy.p0i8.p0i8.i64(i8* nonnull align 32 %t.sroa.0.0.sroa_cast13, i8* align 1 %3, i64 32, i1 false)
   tail call void @llvm.memcpy.p0i8.p0i8.i64(i8* align 1 %3, i8* align 1 %4, i64 32, i1 false)
-  %t.0..sroa_cast10 = bitcast i8* %4 to <4 x i64>*
-  store <4 x i64> %t.0.copyload, <4 x i64>* %t.0..sroa_cast10, align 1
+  call void @llvm.memcpy.p0i8.p0i8.i64(i8* align 1 %4, i8* nonnull align 32 %t.sroa.0.0.sroa_cast13, i64 32, i1 false)
+  call void @llvm.lifetime.end.p0i8(i64 32, i8* nonnull %t.sroa.0.0.sroa_cast13)
   %5 = add i64 %2, 32
   %6 = icmp ugt i64 %5, %len
   br i1 %6, label %bb3, label %bb4
 
-bb11:                                             ; preds = %bb3
-  %t1.0.sroa_cast14 = bitcast %"swap_nonoverlapping_bytes_old::UnalignedBlock"* %t1 to i8*
-  call void @llvm.lifetime.start.p0i8(i64 32, i8* nonnull %t1.0.sroa_cast14)
+bb12:                                             ; preds = %bb3
+  %t1.sroa.0.0.sroa_cast20 = bitcast [4 x i64]* %t1.sroa.0 to i8*
+  call void @llvm.lifetime.start.p0i8(i64 32, i8* nonnull %t1.sroa.0.0.sroa_cast20)
   %7 = sub i64 %len, %i.0.lcssa
   %8 = getelementptr inbounds i8, i8* %x, i64 %i.0.lcssa
   %9 = getelementptr inbounds i8, i8* %y, i64 %i.0.lcssa
-  call void @llvm.memcpy.p0i8.p0i8.i64(i8* nonnull align 8 %t1.0.sroa_cast14, i8* align 1 %8, i64 %7, i1 false)
+  call void @llvm.memcpy.p0i8.p0i8.i64(i8* nonnull align 8 %t1.sroa.0.0.sroa_cast20, i8* align 1 %8, i64 %7, i1 false)
   tail call void @llvm.memcpy.p0i8.p0i8.i64(i8* align 1 %8, i8* align 1 %9, i64 %7, i1 false)
-  call void @llvm.memcpy.p0i8.p0i8.i64(i8* align 1 %9, i8* nonnull align 8 %t1.0.sroa_cast14, i64 %7, i1 false)
-  call void @llvm.lifetime.end.p0i8(i64 32, i8* nonnull %t1.0.sroa_cast14)
-  br label %bb18
+  call void @llvm.memcpy.p0i8.p0i8.i64(i8* align 1 %9, i8* nonnull align 8 %t1.sroa.0.0.sroa_cast20, i64 %7, i1 false)
+  call void @llvm.lifetime.end.p0i8(i64 32, i8* nonnull %t1.sroa.0.0.sroa_cast20)
+  br label %bb20
 
-bb18:                                             ; preds = %bb11, %bb3
+bb20:                                             ; preds = %bb12, %bb3
   ret void
 }
```

Does anyone see an obvious slowdown here? @rkruppe @rust-lang/wg-compiler-nll would be nice to have an LLVM expert look at this.

**Comment by gnzlbg** [comment]
The LLVM-IR generated for `%"MaybeUninit<swap_nonoverlapping_bytes::Block>" = type { [4 x i64] }` has the problem: 

* IR generated by rust -C opt-level=0: https://rust.godbolt.org/z/N-BmJb
* passing that IR through opt: https://gcc.godbolt.org/z/JU4YON
* passing opt output through llc: https://gcc.godbolt.org/z/VE6XDN

generates: 

```asm
        vmovups ymm0, ymmword ptr [rbx + rax]
        vmovaps ymmword ptr [rsp], ymm0
        vmovups ymm0, ymmword ptr [r15 + rax]
        vmovups ymmword ptr [rbx + rax], ymm0
        vmovaps ymm0, ymmword ptr [rsp]
        vmovups ymmword ptr [r15 + rax], ymm0
```

but if we generate `<4 x i64>` instead then we generate what the original code generates:

* IR passed to opt: https://gcc.godbolt.org/z/gcE9vg
* llc machine code: https://gcc.godbolt.org/z/Ry6irk

```asm
        vmovups ymm0, ymmword ptr [rbx + rax]
        vmovups ymm1, ymmword ptr [r15 + rax]
        vmovups ymmword ptr [rbx + rax], ymm1
        vmovups ymmword ptr [r15 + rax], ymm0
```

**Comment by gnzlbg** [comment]
There is also something going on that I cannot figure out. 

If instead of 

```rust
let mut t = mem::MaybeUninit::<Block>::uninitialized();
```

we use 

```rust
let mut t: mem::MaybeUninit<Block> = mem::MaybeUninit { value: mem::uninitialized() };
```

then we generate the same code as before, but if we use either of the following we do not: 

```rust
let mut t: mem::MaybeUninit<Block> = mem::MaybeUninit { uninit: () };
let mut t: mem::MaybeUninit<Block> = mem::uninitialized();
```

we do not =/

**Comment by RalfJung** [comment]
I removed the `ptr` changes; locally that fixes the regression.

@bors try

**Comment by RalfJung** [comment]
@gnzlbg Thanks for looking into this! So it's about `[4 x i64]` vs `<4 x i64>`? What to the angle brackets even mean?

Also I am confused by your 2nd post. First you say `mem::MaybeUninit { value: mem::uninitialized() }` generates the original code again, but then you say it is vastly different from `mem::uninitialized()` which the original code used? Or is it about `mem::uninitialized()` once being at type `T` and once at `MaybeUninit<T>`? Maybe we have to tweak layout computation for unions to compute a layout closer to `T` (but not inheriting any of its restrictions in terms of allowed values)?

**Comment by gnzlbg** [comment]
> What to the angle brackets even mean?

`[ ]` is an array, and `< >` is a vector type.

> First you say mem::MaybeUninit { value: mem::uninitialized() } generates the original code again, but then you say it is vastly different from mem::uninitialized() which the original code used?

These two generate different code:

```rust
let mut t: mem::MaybeUninit<Block> 
    = mem::MaybeUninit { value: mem::uninitialized() }; // ok
let mut t: mem::MaybeUninit<Block> = mem::uninitialized();  // regression
```
I'll try to investigate a bit more about this later and fill an issue with an MWE.

**Comment by RalfJung** [comment]
@gnzlbg I see. What does
```
let mut t: mem::MaybeUninit<Block> = {
  let t: Block = mem::uninitialized();
  mem::transmute(t)
};
```
do? I assume that's also fast. It seems to be about the type at which we call `mem::uninitliazed()`.

But it seems strange that even literally not initializing it at all also makes it slow.

**Comment by kennytm** [comment]
@rust-timer build 9177334470b1f7a7da727bba9ebc4db9d3188e2e

**Comment by RalfJung** [comment]
Uh, timer will work even though bors timed out?

**Comment by pietroalbini** [comment]
[The try build actually passed](https://travis-ci.org/rust-lang/rust/builds/439605201), seems like bors didn't receive the webhook.

**Comment by RalfJung** [comment]
Seems good to me (the red ones have a question mark).

**Comment by RalfJung** [comment]
@cramertj This does not have a perf regression any more. Are you okay with landing this, or should we first figure out how to fix whatever impact `MaybeUninit` has on codegen?

**Comment by gnzlbg** [comment]
cc @eddyb this might be a subtle footgun, is there something that can be done about it ?

**Comment by TimNN** [comment]
Ping from triage! What is the status of this PR?

**Comment by RalfJung** [comment]
This PR needs someone with LLVM knowledge to figure out where the regression is coming from. (In the current state, the PR doesn't regress any more, but that's because I reverted the most used functions affected by this -- there might be codegen issues even with this PR that just don't affect rustc.)

This PR is also basically a soundness fix, we want to get rid of `mem::uninitialized` because it is a horrible footgun. We just have to find a way to do that without regressing performance...

**Comment by rkruppe** [comment]
What @dotdash wrote seems very plausible to me. I do not expect that it will be fixed in LLVM -- arrays generally should not be SSA values, and converting from array to vector to get an SSA value seems very fiddly and rarely-needed.

As for a fix in rustc: while unions clearly can't inherit most representation properties from their contents (e.g., niches), tweaking the ABI to be `Abi::Vector` instead of `Abi::Aggregate` sometimes (e.g., if one of the members is `Abi::Vector`) seems perfectly fine? @eddyb would be the person to ask, though.

**Comment by RalfJung** [comment]
> As for a fix in rustc: while unions clearly can't inherit most representation properties from their contents (e.g., niches), tweaking the ABI to be Abi::Vector instead of Abi::Aggregate sometimes (e.g., if one of the members is Abi::Vector) seems perfectly fine? 

That's basically what I was hoping for. This might take the form of a `repr(transparent)` on unions, where all but one variant must be ZST, and then we'll take the ABI *but not the niche* of that one variant.

**Comment by eddyb** [comment]
I think we can pretty much just copy the ABI of the only non-ZST union field, yeah.
I can mentor someone on this, on IRC/Discord.

**Comment by eddyb** [comment]
@ogoffart That's the right part of the code, but I'm not sure why you're using `Abi::Uninhabited`.
A ZST is `Abi::Aggregate { sized: true }` with `size` 0.
The `abi != field.abi` check seems correct, otherwise.

**Comment by ogoffart** [comment]
@eddyb thanks for the comment. I indeed confused zero-sized and unhinabited. I made a new patch https://github.com/ogoffart/rust/commit/30854cf6eb48b4a6dcf191ca35b98b63636c1a2e  
But how can i test it? where should i put a test for this? I would like to make the test before i can make a pull request.  

I don't understand the comment about Scalar and ScalarPair.  If the ranges are different, the two abi will be different and so we'll fallback to Aggregate.  Or you are saying I should optimize by merging the ranges?

**Comment by Mark-Simulacrum** [comment]
@RalfJung Can you update the current status of this PR for triage? It looks like discussion has died down and it's not super clear if this still needs to land in the current state; I believe some subset of it has already landed though.

**Comment by RalfJung** [comment]
This is currently blocked on https://github.com/rust-lang/rust/pull/55834. Once that lands, we'll do another perf run.

**Comment by RalfJung** [comment]
I can reproduce this locally:
```
Building stage1 std artifacts (x86_64-unknown-linux-gnu -> x86_64-unknown-linux-gnu)
   Compiling core v0.0.0 (/home/r/src/rust/rustc.3/src/libcore)                                                                                                                                                    
   Compiling build_helper v0.1.0 (/home/r/src/rust/rustc.3/src/build_helper)                                                                                                                                       
   Compiling unwind v0.0.0 (/home/r/src/rust/rustc.3/src/libunwind)                                                                                                                                                
   Compiling compiler_builtins v0.0.0 (/home/r/src/rust/rustc.3/src/rustc/compiler_builtins_shim)                                                                                                                  
   Compiling rustc_msan v0.0.0 (/home/r/src/rust/rustc.3/src/librustc_msan)                                                                                                                                        
   Compiling rustc_tsan v0.0.0 (/home/r/src/rust/rustc.3/src/librustc_tsan)                                                                                                                                        
   Compiling rustc_asan v0.0.0 (/home/r/src/rust/rustc.3/src/librustc_asan)                                                                                                                                        
   Compiling rustc_lsan v0.0.0 (/home/r/src/rust/rustc.3/src/librustc_lsan)                                                                                                                                        
   Compiling std v0.0.0 (/home/r/src/rust/rustc.3/src/libstd)                                                                                                                                                      
   Compiling libc v0.0.0 (/home/r/src/rust/rustc.3/src/rustc/libc_shim)                                                                                                                                            
   Compiling alloc v0.0.0 (/home/r/src/rust/rustc.3/src/liballoc)                                                                                                                                                  
   Compiling panic_abort v0.0.0 (/home/r/src/rust/rustc.3/src/libpanic_abort)                                                                                                                                      
   Compiling panic_unwind v0.0.0 (/home/r/src/rust/rustc.3/src/libpanic_unwind)                                                                                                                                    
rustc: /home/r/src/rust/rustc.3/src/llvm/lib/IR/Instructions.cpp:2423: static llvm::CastInst* llvm::CastInst::Create(llvm::Instruction::CastOps, llvm::Value*, llvm::Type*, const llvm::Twine&, llvm::Instruction*): Assertion `castIsValid(op, S, Ty) && "Invalid cast!"' failed.
error: Could not compile `std`.                                                                                                                                                                                    
```

@eddyb @ogoffart Probably something wrong with https://github.com/rust-lang/rust/pull/55834?

**Comment by eddyb** [comment]
Looks more like we're not set up to handle these unions in codegen.
You should be able to use `RUST_LOG=info,rustc_codegen_ssa::mir` to figure out in what function it was, and maybe the exact MIR that caused it.

**Comment by ogoffart** [comment]
Looks like it is caused by MaybeUninit<bool> for example.  The llvm IR is

```
define internal zeroext i1 @"_ZN40_$LT$core..mem..MaybeUninit$LT$T$GT$$GT$10into_inner17hd46e8fc6d910c21aE"(i8 %self) unnamed_addr #1 {
start:
 %0 = bitcast i8 %self to i1
...
```

And apparently one cannot bitcast i8 to i1

I believe this is because the union is a Scalar with valid_range: 0..=255,  while bool's valid_range is 0..=1

**Comment by gnzlbg** [comment]
Shouldn't that be a `trunc` instead of a `bitcast` ?

**Comment by eddyb** [comment]
But what's actually generating that on the rustc side? Why is it any different from `(bool,)`?

**Comment by ogoffart** [comment]
@eddyb the difference between `union { bool }` and `(bool,)`  is that the union resets the valid_range. (Remember that #55834 does scalar_unit because
https://github.com/rust-lang/rust/pull/54668#issuecomment-437172386 )
I'm not sure this is related, but it seems to be.

**Comment by eddyb** [comment]
@ogoffart Ah, I misread your previous comment, so the `union` looks like an `u8` newtype, but the field is a `bool`. Maybe a `to_immediate` missing somewhere?
Where is this coming from, `OperandRef::extract_field`?

**Comment by ogoffart** [comment]
@eddyb Right, I believe it comes from the bitcast in OperandRef::extract_field after the "HACH" comment.   Note that the match before compare `field.size == self.layout.size` which is equal for i8 and i1, and then will proceed to do a bitcast.   

I don't really know where to_immediate needs to be inserted,  probably not in OperandRef::extract_field itself because of the type difference.  
Maybe OperandRef::extract_field should have a special case for bool ?

**Comment by eddyb** [comment]
@ogoffart Okay I've taken a closer look and this is not as bad as I thought:
You can grab the `Scalar`s for immediates and pairs, like here:
https://github.com/rust-lang/rust/blob/046e054a995c7b293e26968774ff88b38661cd3c/src/librustc_codegen_ssa/mir/operand.rs#L90-L118
(note that for `OperandValue::Immediate` you shouldn't panic if you don't find a `Scalar` - it just means that you can skip any casting)

Then you can do `to_immediate_scalar` *before* the `bitcast`s (you can even replace the `bitcast`s with `pointercast`s if you check that `scalar.value` is `Pointer`).

Alternatively, you could check that the type passed to `bitcast` is `== bx.cx().type_i1()` and do `trunc` instead of `bitcast`.

cc @rkruppe @nagisa

**Comment by RalfJung** [comment]
I merged #56097 into this so that we can do a perf test.

@bors try

**Comment by kennytm** [comment]
@rust-timer build 979ef5e85580f3a9a53ed504dba24ac4c8682019

**Comment by RalfJung** [comment]
Ouch, that is still very red...

**Comment by ogoffart** [comment]
Perhaps the issue is anyway related to the functions not being inlined. Adding ` #[inline(always)]` to the  MaybeUninit functions could help.  (maybe also ManuallyDrop)

**Comment by RalfJung** [comment]
I seriously doubt LLVM wouldn't inline this, but let's see.

@bors try

**Comment by varkor** [comment]
@rust-timer build 764c4682552ecdbd88b31e7de54347c040162f87

**Comment by RalfJung** [comment]
Uh, the try job isn't done yet...

**Comment by varkor** [comment]
[Apparently it works even before](https://discordapp.com/channels/442252698964721669/443151243398086667/513911955942277140) the try run is finished 😁

**Comment by RalfJung** [comment]
Please don't use links to discord, people without an account cannot read them. Actually not even I can just follow the link despite me being logged in (it asks me to log in again?!?).

**Comment by shepmaster** [comment]
> Please don't use links to discord,

Or at least copy-paste the important details, providing the link for further citation.

**Comment by RalfJung** [comment]
Oh wow that looks *much* better, thanks for the suggestion @ogoffart!

@rust-lang/wg-compiler-performance is it expected that https://github.com/rust-lang/rust/pull/54668/commits/7b9d1394a59288c3a1f29f66628da441d225f000 makes such a huge difference in perf?

Now, we still have a 2.8% regression in one benchmark, but we also have a 2.1% improvement elsewhere. So this looks fairly good to me?

**Comment by rkruppe** [comment]
I don't think the `always` part is at all relevant, but it's known that `#[inline]` can heavily enable cross crate inlining even on generic functions in the default configuration (multiple codegen units, crate-local ThinLTO, no cross-crate LTO). See e.g. discussion in https://github.com/rust-lang/rust/pull/52704

**Comment by rkruppe** [comment]
In fact I would advocate for another perf run with only `#[inline]` annotations in those places, no `inline(always)`.

**Comment by RalfJung** [comment]
> In fact I would advocate for another perf run with only #[inline] annotations in those places, no inline(always).

All right.

@bors try

**Comment by nnethercote** [comment]
> @rust-lang/wg-compiler-performance is it expected that [7b9d139](https://github.com/rust-lang/rust/commit/7b9d1394a59288c3a1f29f66628da441d225f000) makes such a huge difference in perf?

Inlining of a hot function can certainly make a noticeable difference, and I've landed multiple PRs doing just that to win 1% or 2% on some benchmarks. The size of the effect on `tokio-webpush-simple` -- over 30% on one case! -- is surprising, though.

> Now, we still have a 2.8% regression in one benchmark, but we also have a 2.1% improvement elsewhere. So this looks fairly good to me?

Looks good to me. I suspect the 2.1% improvement might be noise, because it's for `style-servo-opt` but `style-servo-check` and `style-servo-debug` didn't improve. And the regression for `issue-46449` doesn't really matter because it's not an important benchmark.

**Comment by kennytm** [comment]
@rust-timer build 879a3bbb98935798f33c4615aa074735f4bf9bcb

**Comment by nnethercote** [comment]
New results are a bit worse. Many people seem to not like `inline(always)`, but I think it's ideal for cases like this where the inlining is known to be crucial. Especially when the functions are so small.

**Comment by dotdash** [comment]
> Since the benchmark compute the compilation time, it might be that giving hint about how to inline makes the compilation faster.

Wouldn't that only make sense when the change was in the code that is being compiled?

**Comment by ogoffart** [comment]
> Wouldn't that only make sense when the change was in the code that is being compiled?

Yes.
Since the change is within some basic inline functions within the core library, I guess it is likely to be part of the code that is being compiled.

**Comment by eddyb** [comment]
@ogoffart Keep in mind that if this is happening in `cargo check` testcases, all you'll be seeing is how fast or slow `rustc` is, with LLVM not being involed *at all*.
And all of these libcore changes *definitely can* impact `rustc`.

File: src/libcore/ptr.rs
Comment (RalfJung) [reply src/libcore/ptr.rs]: I added back a comment about `drop`.

**Comment by RalfJung** [comment]
I made the *really* trivial functions `inline(always)` now.

@bors try

**Comment by ogoffart** [comment]
> Keep in mind that if this is happening in cargo check testcases, all you'll be seeing is how fast or slow rustc is, with LLVM not being involed at all.

If I read the number correctly, the "-check" benchmark does not seem to be impacted. The red numbers are only in the "-opt" and "-debug" benchmarks

**Comment by RalfJung** [comment]
Unions got fixed, I rebased this PR.

So, can we land this?

**Review by ogoffart — APPROVED** [review]
(I'm in no position to review, but i see no problem with this, and since it is required to stabilize MaybeUnitit, it'd be great to have it in as early as possible)

**Comment by gnzlbg** [comment]
Finally, awesome, thanks!

**Pull request merged**
